gc.c

```diff
@@ -957,6 +957,110 @@ static const char *obj_info(VALUE obj);
  * http://www.mcs.anl.gov/~kazutomo/rdtsc.html
  * written by Kazutomo Yoshii <kazutomo@mcs.anl.gov>
  */
+/* Source code to use RDTSC is modified to serialize
+ * RDTSC, it is unordered on x86 processors and the results
+ * can be non-monotonic if compared on different processors.
+ * https://lkml.org/lkml/2011/5/10/297
+ * modified by https://github.com/ruby/ruby/pull/2330
+ */
+
+#if defined(__i386__) || defined(__x86_64__)
+#if defined(__GNUC__)
+
+#if defined(__i386__)
+static int
+has_cpuid()
+{
+    static volatile int initialized = 0;
+    static uint32_t flags;
+    if (initialized)
+        return (flags & 0x200000) != 0;
+    /* http://wiki.osdev.org/CPUID#Checking_CPUID_availability */
+    __asm__ __volatile__ (
+        "pushf{l|d}\n\t" /* Save EFLAGS */
+        "pushf{l|d}\n\t" /* Store EFLAGS */
+        "xor{l $0x00200000, (%%esp)| dword ptr [esp], 0x00200000}\n\t" /* Invert the ID bit in stored EFLAGS */
+        "popf{l|d}\n\t" /* Load stored EFLAGS (with ID bit inverted) */
+        "pushf{l|d}\n\t" /* Store EFLAGS again (ID bit may or may not be inverted) */
+        "pop {%%}eax\n\t" /* eax = modified EFLAGS (ID bit may or may not be inverted) */
+        "xor {(%%esp), %%eax|eax, [esp]}\n\t" /* eax = whichever bits were changed */
+        "popf{l|d}\n\t" /* Restore original EFLAGS */
+        : "=a" (flags));
+    initialized = 1;
+    return (flags & 0x200000) != 0;
+}
+#endif
+
+static inline void
+cpuid(uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx)
+{
+    __asm__ __volatile__ (
+#if defined(__i386__) /* This is needed for -fPIC to work on i386 */
+    "movl %%ebx, %%esi\n\t"
+#endif
+    "cpuid\n\t"
+#if defined(__i386__)
+    "xchgl %%ebx, %%esi\n\t"
+    : "=a" (*eax), "=S" (*ebx), "=c" (*ecx), "=d" (*edx) : "a" (*eax));
+#else
+    : "=a" (*eax), "=b" (*ebx), "=c" (*ecx), "=d" (*edx) : "a" (*eax));
```

Comment:
Writing closing parentheses in both side of `#if` confuses editors.
Selecting whole `__asm__` statements seems same in numbers of lines.